WIP FORMS-202 # BMP is sending down hidden fields from the server with an

simonmarklar opened the pull request:
attribute of 'hide'. This is now accounted for.

jokeyrhyme commented:
I'm a little ambivalent about how different this is to the Forms v2 implementation.
I think it might be better for performance to remove any Elements hidden by a SubFormElement's overrides completely from the definition, early in the process. That way we avoid creating Backbone models, listeners, etc.

Also, Forms v2 implements other sorts of SubFormElement overrides, e.g. changing the available options in a choice Element. We don't need to implement this right away, but we should at least account for it and see if the hidden work here can be done in a way that allows for other overrides to be implemented consistently in future. If this turns out to be too troublesome, then that's fine, but I'd like to double-check that this was considered.

I'm also wondering whether the JSON definition generated on the server might need to be modified if we plan to implement all the other sorts of overrides...

simonmarklar commented:
it would definitely be better to remove it from the definition and not send it down to the client. but that requires a platform update, right ? and if thats the case, why were they marked with 'hide' in the first place ?

 I was looking at the form definition sent down, it included so many unrelated forms that I didnt think should be in the definition for the form i was testing. But im just guessing there, perhaps there is a reason for the info sent down?

as it stands, there are fields that are marked as hidden that are being shown. This fix hides the elements, which also removes them from the validation. This makes it look to the user that it functions like it used to in BIC2, and that the hide functionality isnt broken, while we can check out removing them from the definition

jokeyrhyme commented:
It can't be removed from the definition by the server.
This functionality can only be applied once both the parent form and the child form are used together, because it is information on the parent form that overrides the behaviour of the child form (with this feature).

But, it _can_ be removed from the child's definition after it is retrieved from storage and before it is used.

The JSON definition structure for this setting probably needs to be reviewed, as I'm not sure it is being migrated correctly.

simonmarklar commented:
I'm not sure I understand, why couldnt the server build the definition and exclude fields based on the hidden flag??? its just data, we can transform it however we want.

Given the current focus on performance and slow form loading times I would have thought that it make more sense to exclude the fields on the server and send down a smaller definition, rather than iterate over all the fields and remove them at the time of getting the definition.

anyways, ill find some way of pruning out the fields after they've been downloaded, but i still think they shouldn't even come down

jokeyrhyme commented:
The design for Forms includes:

- start with a definition for the parent form
- that parent form may include a sub-form element, which is associated with a child form definition
- that sub-form element may link to multiple instances of that definition (a.k.a. sub-records)

The sub-form override feature is designed as follows:
- you have a child form definition that is almost exactly what you want to use with your parent form
- it's not perfect though, and you wish you didn't have to duplicate the whole child form just to change a few little things (we also want to minimise our use of device storage when possible)
- so this feature allows you to declare as properties on the sub-form element (the link between parent and child) how the child should be changed when part of this specific parent
- multiple parent forms may re-use the same shared child form definition, and each parent form may override different parts of the child form definition

This is why the server cannot know statically how to modify the child form definitions. It is only at run-time that the parent and it's overrides over the child are known.

ashishtilara commented:
builder should attach `_elements` object to subform field definition for overrides, I'll try and generate the form and see if that is happening correctly.

ashishtilara commented:
okay, confirmed, it is the case... however I noticed, builder is not turning `options` to object, its a string...

jokeyrhyme commented:
So, maybe we need to split the JIRA Issue into both front-end and back-end components? To address the options String problem? Meanwhile, this front-end Issue can optionally parse the String, otherwise treat it like an Object?

simonmarklar commented:
Yes, the  _elements property was what i was getting at. rather than having the client go through each element in `_elements` looking for a `hide:1` and then remove that element from the array, why cant it just not be included in the array in the first place? removing a hidden field from `_elements` in a test definition stops the field from being created

each sub form element defines which fields are hidden, and this is done at in the builder and reflected in the `_element` property - so unless im missing something, everything is known by the server.

jokeyrhyme commented:
Everything _is_ known by the server, but not during use of the Forms Builder. The Forms Builder does not open up all parents of a child form, and cannot know how to permanently and globally alter the child form to satisfy all parent forms. Also, only overrides that were common could be processed like this (e.g. all parent form specify that they don't need "element1" from the shared child form).

As soon as any two parent forms have different expectations of a shared child form, the only way the server can reconcile this would be to modify a duplicate of the child form, at least once per disagreement. This defeats the purpose of the feature in question.

If there is only 1 parent form, or all parent forms have the same overrides over their shared child form, then the server could optimise this by modifying the child form as you state, @simonmarklar. However, if this was the case, then the humans using the Forms Builder would be better off not using the feature in question at all, as it wouldn't be necessary.

jokeyrhyme commented:
After thinking about some of the existing use cases we have, I think the main stories addressed by the implementation of this feature in Forms v2 are:

- When multiple form definitions have very few differences, it should be possible to only transfer the common parts once along with the differences, so that devices don't waste storage space on duplicate definitions.

- When working with repeat definition structures, it would be more convenient for a forms designer to manage differences at the parent form level, as the Forms Builder makes tweaking large numbers of child forms very tedious.

I'll cede that reimplementing this particular Forms v2 feature is _not_ the only way to address these needs.

simonmarklar commented:
im not suggesting modifying the child forms  - im suggesting not including the elements defined as  hidden in the** subform field**'s _element property. This is on a per form basis and doesnt touch the actual sub form.

I'll see if i can illustrate what im talking about :) 

here's a form with a subform definition:
```
{
    "default": {
        "uniqueNameId": "22uehx",
        "name": "firstLevel",
        "formDescription": "",
        "defaultCategory": "",
        "maxStep": 6,
        "labelPlacement": "auto",
        "header": "",
        "footer": "",
        "_elements": [{
            "default": {
                "name": "id",
                "type": "text",
                "page": 0
            }
        }, {
            "default": {
                "name": "first_level_req",
                "type": "text",
                "label": "First & Level' \" Req ",
                "labelPlacement": "default",
                "labelStyle": "Plain",
                "required": "1",
                "maxWidthPrefix": "characters",
                "page": 0
            }
        }, {
            "default": {
                "name": "second_level_field",
                "type": "subForm",
                "subForm": "second_level_form",
                "_elements": {
                    "second|required": {
                        "override": "",
                        "type": "textbox",
                        "id": "second_required",
                        "hide": "1"
                    },
                    "nothidden": {
                        "override": "",
                        "type": "textbox",
                        "id": "nothidden",
                        "hide": "1"
                    },
                    "third|level|field": {
                        "hide": "",
                        "override": "",
                        "type": "sub_form",
                        "id": "third_level_field"
                    },
                    "cond|hidden": {
                        "hide": "",
                        "override": "",
                        "type": "textbox",
                        "id": "cond_hidden"
                    }
                },
                "labelPlacement": "default",
                "labelStyle": "Plain",
                "subformControlPos": "below",
                "subformPerms": "allow_add",
                "preload": "no",
                "page": 0
            }
        }],
        "_checks": [],
        "_actions": [],
        "_behaviours": []
    },
    "list": {
        "interaction": "FIRSTLEVEL_LIST",
        "displayName": "FIRSTLEVEL LIST",
        "defaultCategory": "",
        "header": "",
        "footer": "",
        "hidden": null,
        "action": "list",
        "orderBy": null,
        "showLink": {
            "edit": "1",
            "view": "1",
            "delete": "1"
        },
        "dataStorageList": "temporary",
        "downloadModeList": "partial",
        "recordsToDisplay": "",
        "_elements": ["first_level_req", "second_level_field"]
    },
    "search": {
        "interaction": null,
        "displayName": null,
        "defaultCategory": null,
        "header": null,
        "footer": null,
        "hidden": null,
        "action": "search"
    },
    "add": {
        "interaction": "FIRSTLEVEL_ADD",
        "displayName": "FIRSTLEVEL ADD",
        "defaultCategory": "",
        "header": "",
        "footer": "",
        "hidden": null,
        "action": "add",
        "_elements": ["first_level_req", "second_level_field"]
    },
    "edit": {
        "interaction": "FIRSTLEVEL_EDIT",
        "displayName": "FIRSTLEVEL EDIT",
        "defaultCategory": "",
        "header": "",
        "footer": "",
        "hidden": "1",
        "action": "edit",
        "_elements": ["first_level_req", "second_level_field"]
    },
    "view": {
        "interaction": "FIRSTLEVEL_VIEW",
        "displayName": "FIRSTLEVEL VIEW",
        "defaultCategory": "",
        "header": "",
        "footer": "",
        "hidden": "1",
        "action": "view",
        "_elements": ["first_level_req", "second_level_field"]
    },
    "delete": {
        "interaction": "FIRSTLEVEL_DELETE",
        "displayName": "FIRSTLEVEL DELETE",
        "defaultCategory": "",
        "header": "",
        "footer": "",
        "hidden": "1",
        "action": "delete",
        "_elements": ["first_level_req", "second_level_field"]
    }
}
```

see the object with a name of `second_level_field` ? it has a `_elements` property.  if you take out any of the elements with `hide:1` set in it, the current client will render the subform without the element 

If you define a second Form that uses the subform `second_level_form`, it will have its own subform **element**, with its own `_elements` list of overrides. 

Again, i'm not saying change the subform - im saying change the `_elements` property on the **subform element** - the field that holds the subform and provides the overrides.

jokeyrhyme commented:
Ah, okay. Yes, I didn't realise this was the structure we were exporting as JSON when this feature was in use. Everything you've said here, @simonmarklar, makes sense to me now. I should actually research my opinions in future. :P

So, in light of the current JSON structure, I'm actually kinda' happy with the structure.

When interpreting the child form definition, it's the `"_elements"` lists in the `"add"`, `"edit"`, etc variations that are currently used to efficiently eliminate certain elements form the child form.

If we want to avoid Backbone Model instantiation completely, then the place to do this would be in "forms/models/elements/subform", in the `#add()` method, after `Forms.getDefinition()` and before `SubFormModel.create()`.

simonmarklar commented:
I dont explain it very well either :)

I just want to point out that specifying hiding of subforms fields (at least, the behavior i see in the forms builder) only happens on the `default` option - you cant hide a field in the `add`, `edit` etc tabs. I assume the default flows through to the add, edit etc forms.

Im just testing out what happens when the element is removed but the subform has fields with behaviors that reference the removed field - if that can even happen, so far it seems seems ok.

jokeyrhyme commented:
Behaviours that reference missing fields should complain but not explode. Hopefully. :)

simonmarklar commented:
OK now this is painful.

I wanted to see what happens with behaviors that referenced a removed field. So I made a definition on  blinkm.co with a hidden field and another field that showed itself when the hidden field had a value of 'a', and now im getting completely different behavior from a definition.

The `hide:"1"` value is no longer being applied from the `default._elements`  property - changing any field, or even removing an element from this list has no effect on the form - in fact i'm not seeing any of the overridden properties coming though on the definition. I put logs in `models/elements/subform.js` where it gets the definition and this shows me that the definition being returned does not have any of the overridden fields. 

`Forms.getDefintion` is passed the subform field name, but I actually dont see anywhere that combines that subform **form** with the subform field **`_elements`** list. maybe its in @blinkmobile/varied-definition ??

Comparing the previous definition with the newest definition it looks like in the previous definition I was using has `hide:1` set on the subform definition themselves!! This is really puzzling to me, as both definitions were made on blinkm.co and then retrieved from the server response chrome network panel.  The only place in the Forms builder where you can set the `hide` field is in the Form builders edit section for a subform element. I do notice that the new subform I built doesnt have any interactions where as the old one did (before I added a parent form to a sub form and it blew up :O )

The first form has been my subform "testing" form, which has had things added and removed and its been regenerated quite a few times - I've only ever formatted it because I dont  know the format - I always build and generate from blinkm.co .. perhaps I did something in the builder that added the `hide:1` fields?

Anyways it appears that I was wrong about removing the field from the `_elements` array in a subform field being a way to hide the element - at least with the current definition that I have.

but this does open up another can of worms. namely, why are the overrides of a sub form element field not being applied to a subform model ??

here's my complete definition from my answerspace:

```
[{
   "default": {
     "uniqueNameId": "4yf1xf7xt32",
     "name": "blink2_camera_test",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "camera",
         "type": "file",
         "label": "Camera",
         "capture": true,
         "accept": "image\\\/*",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "BLINK2_CAMERA_TEST_LIST",
     "displayName": "BLINK2_CAMERA_TEST LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["camera"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "BLINK2_CAMERA_TEST_ADD",
     "displayName": "BLINK2_CAMERA_TEST ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["camera"]
   },
   "edit": {
     "interaction": "BLINK2_CAMERA_TEST_EDIT",
     "displayName": "BLINK2_CAMERA_TEST EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["camera"]
   },
   "view": {
     "interaction": "BLINK2_CAMERA_TEST_VIEW",
     "displayName": "BLINK2_CAMERA_TEST VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["camera"]
   },
   "delete": {
     "interaction": "BLINK2_CAMERA_TEST_DELETE",
     "displayName": "BLINK2_CAMERA_TEST DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["camera"]
   }
 }, {
   "default": {
     "uniqueNameId": "us90ex2",
     "name": "every_field_type",
     "formDescription": "a form for testing every field",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "Form header",
     "footer": "Form footer",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "text_area",
         "type": "textarea",
         "label": "Text Area",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "placeholderText": "this field is required",
         "wrap": "hard",
         "page": 0
       }
     }, {
       "default": {
         "name": "password",
         "type": "password",
         "label": "Password",
         "page": 0
       }
     }, {
       "default": {
         "name": "email",
         "type": "email",
         "label": "Email",
         "page": 0
       }
     }, {
       "default": {
         "name": "url",
         "type": "url",
         "label": "Url",
         "page": 0
       }
     }, {
       "default": {
         "name": "phone_number",
         "type": "telephone",
         "label": "Phone Number",
         "page": 0
       }
     }, {
       "default": {
         "name": "number",
         "type": "number",
         "label": "Number",
         "page": 0
       }
     }, {
       "default": {
         "name": "currency",
         "type": "number",
         "label": "Currency",
         "page": 0
       }
     }, {
       "default": {
         "name": "calculation",
         "type": "message",
         "label": "Calculation",
         "persist": true,
         "page": 0
       }
     }, {
       "default": {
         "name": "radio",
         "type": "select",
         "label": "Radio",
         "mode": "expanded",
         "page": 0
       }
     }, {
       "default": {
         "name": "select",
         "type": "select",
         "label": "Select",
         "mode": "collapsed",
         "page": 0
       }
     }, {
       "default": {
         "name": "star_rating",
         "type": "number",
         "label": "Star Rating",
         "mode": "graphical",
         "graphic": {
           "active": {
             "ascii": "*"
           },
           "inactive": {
             "ascii": "*"
           }
         },
         "min": "0",
         "max": "5",
         "page": 0
       }
     }, {
       "default": {
         "name": "multi",
         "type": "multi",
         "label": "Multi",
         "mode": "collapsed",
         "page": 0
       }
     }, {
       "default": {
         "name": "checkboxes",
         "type": "multi",
         "label": "Checkboxes",
         "mode": "expanded",
         "page": 0
       }
     }, {
       "default": {
         "name": "checkbox",
         "type": "boolean",
         "label": "Checkbox",
         "options": ["n", "y"],
         "defaultValue": 0,
         "page": 0
       }
     }, {
       "default": {
         "name": "date",
         "type": "date",
         "label": "Date",
         "page": 0
       }
     }, {
       "default": {
         "name": "time",
         "type": "time",
         "label": "Time",
         "page": 0
       }
     }, {
       "default": {
         "name": "timestamp",
         "type": "datetime",
         "label": "Timestamp",
         "page": 0
       }
     }, {
       "default": {
         "name": "camera",
         "type": "file",
         "label": "Camera",
         "capture": true,
         "accept": "image\\\/*",
         "page": 0
       }
     }, {
       "default": {
         "name": "image_library",
         "type": "file",
         "label": "Image Library",
         "accept": "image\\\/*",
         "page": 0
       }
     }, {
       "default": {
         "name": "file_upload",
         "type": "file",
         "label": "File Upload",
         "page": 0
       }
     }, {
       "default": {
         "name": "_heading_1",
         "type": "heading",
         "text": "This should be a new page",
         "headingSmallText": "Shazza got us some crook also lets throw a trackies. She'll be right rollie how as stands out like roo bar. Lets get some mullet to flat out like a flake. Lets get some franger to lets throw a rotten. Trent from punchy stickybeak no dramas as stands out like dunny. vb my it'll be dag. Mad as a muster flamin she'll be right deadset. Get a dog up ya kindie where grab us a grouse. As cross as a rock up with she'll be right dill.",
         "headingPageBreak": "1",
         "position": "left",
         "fontFace": "arial",
         "page": 1
       }
     }, {
       "default": {
         "name": "textbox",
         "type": "text",
         "label": "Textbox",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "tooltip": "I am the tool tip for the text area",
         "hint": "I am the hint text for the text area",
         "maxWidthPrefix": "characters",
         "page": 1
       }
     }, {
       "default": {
         "name": "location",
         "type": "location",
         "label": "Location",
         "page": 1
       }
     }, {
       "default": {
         "name": "sketch_signature",
         "type": "draw",
         "label": "Sketch Signature",
         "size": "signature",
         "page": 1
       }
     }, {
       "default": {
         "name": "avalue",
         "type": "text",
         "label": "Avalue",
         "page": 1
       }
     }, {
       "default": {
         "name": "logged_in_user_id",
         "type": "text",
         "label": "Logged In User Id",
         "page": 1
       }
     }, {
       "default": {
         "name": "user_attribute",
         "type": "text",
         "label": "User Attribute",
         "page": 1
       }
     }, {
       "default": {
         "name": "get_value",
         "type": "text",
         "label": "Get Value",
         "page": 1
       }
     }, {
       "default": {
         "name": "subform",
         "type": "subForm",
         "subForm": "firstLevel",
         "page": 1
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "",
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": null,
       "view": null,
       "delete": null
     },
     "dataStorageList": null,
     "downloadModeList": null,
     "recordsToDisplay": null
   },
   "search": {
     "interaction": "",
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "EVERY_FIELD_TYPE_ADD",
     "displayName": "EVERY_FIELD_TYPE ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["text_area", "password", "email", "url", "phone_number", "number", "currency", "calculation", "radio", "select", "star_rating", "multi", "checkboxes", "checkbox", "date", "time", "timestamp", "camera", "image_library", "file_upload", "_heading_1", "textbox", "location", "sketch_signature", "avalue", "logged_in_user_id", "user_attribute", "get_value", "subform"]
   },
   "edit": {
     "interaction": "EVERY_FIELD_TYPE_EDIT",
     "displayName": "EVERY_FIELD_TYPE EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["text_area", "password", "email", "url", "phone_number", "number", "currency", "calculation", "radio", "select", "star_rating", "multi", "checkboxes", "checkbox", "date", "time", "timestamp", "camera", "image_library", "file_upload", "_heading_1", "textbox", "location", "sketch_signature", "avalue", "logged_in_user_id", "user_attribute", "get_value", "subform"]
   },
   "view": {
     "interaction": "",
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "view",
     "_elements": ["subform"]
   },
   "delete": {
     "interaction": "",
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "delete",
     "_elements": ["subform"]
   }
 }, {
   "default": {
     "uniqueNameId": "349lq",
     "name": "filetest",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "upload",
         "type": "file",
         "label": "Upload",
         "page": 0
       }
     }, {
       "default": {
         "name": "text",
         "type": "text",
         "label": "Text",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "FILETEST_LIST",
     "displayName": "FILETEST LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["upload", "text"]
   },
   "search": {
     "interaction": "",
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "FILETEST_ADD",
     "displayName": "FILETEST ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["upload", "text"]
   },
   "edit": {
     "interaction": "FILETEST_EDIT",
     "displayName": "FILETEST EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["upload", "text"]
   },
   "view": {
     "interaction": "FILETEST_VIEW",
     "displayName": "FILETEST VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["upload", "text"]
   },
   "delete": {
     "interaction": "FILETEST_DELETE",
     "displayName": "FILETEST DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["upload", "text"]
   }
 }, {
   "default": {
     "uniqueNameId": "22uehx",
     "name": "firstLevel",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "first_level_req",
         "type": "text",
         "label": "First & Level' \ Req ",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "maxWidthPrefix": "characters",
         "page": 0
       }
     }, {
       "default": {
         "name": "second_level_field",
         "type": "subForm",
         "subForm": "second_level_field",
         "_elements": {
           "second|level|text": {
             "hide": "",
             "override": "",
             "type": "number",
             "id": "second_level_text"
           },
           "hidden|when|first": {
             "override": "",
             "type": "textbox",
             "id": "hidden_when_first",
             "hide": "1"
           },
           "hidden|when|text|is|a": {
             "hide": "",
             "override": "",
             "type": "textbox",
             "id": "hidden_when_text_is_a"
           },
           "third|level": {
             "hide": "",
             "override": "",
             "type": "sub_form",
             "id": "third_level"
           }
         },
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "subformControlPos": "below",
         "subformPerms": "allow_add",
         "preload": "no",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "FIRSTLEVEL_LIST",
     "displayName": "FIRSTLEVEL LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["first_level_req", "second_level_field"]
   },
   "search": {
     "interaction": "",
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "FIRSTLEVEL_ADD",
     "displayName": "FIRSTLEVEL ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["first_level_req", "second_level_field"]
   },
   "edit": {
     "interaction": "FIRSTLEVEL_EDIT",
     "displayName": "FIRSTLEVEL EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["first_level_req", "second_level_field"]
   },
   "view": {
     "interaction": "FIRSTLEVEL_VIEW",
     "displayName": "FIRSTLEVEL VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["first_level_req", "second_level_field"]
   },
   "delete": {
     "interaction": "FIRSTLEVEL_DELETE",
     "displayName": "FIRSTLEVEL DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["first_level_req", "second_level_field"]
   }
 }, {
   "default": {
     "uniqueNameId": "23hax2",
     "name": "forms2_qld_test",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "name",
         "type": "text",
         "label": "Name",
         "page": 0
       }
     }, {
       "default": {
         "name": "address",
         "type": "text",
         "label": "Address",
         "page": 0
       }
     }, {
       "default": {
         "name": "childs_name",
         "type": "text",
         "label": "Childs Name",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "FORMS2_QLD_TEST_LIST",
     "displayName": "FORMS2_QLD_TEST LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["name", "address", "childs_name"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "FORMS2_QLD_TEST_ADD",
     "displayName": "FORMS2_QLD_TEST ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["name", "address", "childs_name"]
   },
   "edit": {
     "interaction": "FORMS2_QLD_TEST_EDIT",
     "displayName": "FORMS2_QLD_TEST EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["name", "address", "childs_name"]
   },
   "view": {
     "interaction": "FORMS2_QLD_TEST_VIEW",
     "displayName": "FORMS2_QLD_TEST VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["name", "address", "childs_name"]
   },
   "delete": {
     "interaction": "FORMS2_QLD_TEST_DELETE",
     "displayName": "FORMS2_QLD_TEST DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["name", "address", "childs_name"]
   }
 }, {
   "default": {
     "uniqueNameId": "b3gcb7pimd",
     "name": "hidden_fields",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "TextField",
         "type": "text",
         "label": "Text Field",
         "page": 0
       }
     }, {
       "default": {
         "name": "NumberField",
         "type": "number",
         "label": "Number Field",
         "page": 0
       }
     }, {
       "default": {
         "name": "subform2",
         "type": "subForm",
         "subForm": "second_level_form",
         "_elements": {
           "second|required": {
             "hide": "",
             "override": "",
             "type": "textbox",
             "id": "second_required"
           },
           "nothidden": {
             "hide": "",
             "override": "",
             "type": "textbox",
             "id": "nothidden"
           },
           "third|level|field": {
             "override": "",
             "type": "sub_form",
             "id": "third_level_field",
             "hide": "1"
           },
           "cond|hidden": {
             "hide": "",
             "override": "",
             "type": "textbox",
             "id": "cond_hidden"
           }
         },
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "subformControlPos": "below",
         "subformPerms": "allow_add",
         "preload": "no",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "HIDDEN_FIELDS_LIST",
     "displayName": "HIDDEN_FIELDS LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["TextField", "NumberField", "subform2"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "HIDDEN_FIELDS_ADD",
     "displayName": "HIDDEN_FIELDS ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["TextField", "NumberField", "subform2"]
   },
   "edit": {
     "interaction": "HIDDEN_FIELDS_EDIT",
     "displayName": "HIDDEN_FIELDS EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["TextField", "NumberField", "subform2"]
   },
   "view": {
     "interaction": "HIDDEN_FIELDS_VIEW",
     "displayName": "HIDDEN_FIELDS VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["TextField", "NumberField", "subform2"]
   },
   "delete": {
     "interaction": "HIDDEN_FIELDS_DELETE",
     "displayName": "HIDDEN_FIELDS DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["TextField", "NumberField", "subform2"]
   }
 }, {
   "default": {
     "uniqueNameId": "4aim",
     "name": "outer_subform",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "outer_name",
         "type": "text",
         "label": "Outer Name",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "maxWidth": "2",
         "maxWidthPrefix": "characters",
         "page": 0
       },
       "list": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "maxWidthPrefix": "characters"
       },
       "add": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "maxWidthPrefix": "characters"
       },
       "edit": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "maxWidthPrefix": "characters"
       },
       "view": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "maxWidthPrefix": "characters"
       },
       "delete": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "maxWidthPrefix": "characters"
       }
     }, {
       "default": {
         "name": "subform",
         "type": "text",
         "label": "Subform",
         "page": 0
       }
     }, {
       "default": {
         "name": "file_test",
         "type": "subForm",
         "subForm": "filetest",
         "_elements": {
           "upload": {
             "hide": "",
             "override": "",
             "type": "file_upload",
             "id": "upload"
           },
           "text": {
             "hide": "",
             "override": "",
             "type": "textbox",
             "id": "text"
           }
         },
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "subformControlPos": "below",
         "maxSubforms": "10",
         "minSubforms": "2",
         "subformPerms": "allow_add",
         "preload": "no",
         "page": 0
       },
       "list": {
         "labelPlacement": "default",
         "labelStyle": "Plain"
       },
       "add": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "subformControlPos": "below"
       },
       "edit": {
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "subformControlPos": "below"
       },
       "view": {
         "labelPlacement": "default",
         "labelStyle": "Plain"
       },
       "delete": {
         "labelPlacement": "default",
         "labelStyle": "Plain"
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "OUTER_SUBFORM_LIST",
     "displayName": "OUTER_SUBFORM LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["outer_name", "subform", "file_test"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "OUTER_SUBFORM_ADD",
     "displayName": "OUTER_SUBFORM ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["outer_name", "subform", "file_test"]
   },
   "edit": {
     "interaction": "OUTER_SUBFORM_EDIT",
     "displayName": "OUTER_SUBFORM EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["outer_name", "subform", "file_test"]
   },
   "view": {
     "interaction": "OUTER_SUBFORM_VIEW",
     "displayName": "OUTER_SUBFORM VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["outer_name", "subform", "file_test"]
   },
   "delete": {
     "interaction": "OUTER_SUBFORM_DELETE",
     "displayName": "OUTER_SUBFORM DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["outer_name", "subform", "file_test"]
   }
 }, {
   "default": {
     "uniqueNameId": "418tzdh0w6md",
     "name": "second_level_field",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 4,
     "labelPlacement": null,
     "header": null,
     "footer": null,
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "second_level_text",
         "type": "text",
         "label": "Second Level Text",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "maxWidthPrefix": "characters",
         "page": 0
       }
     }, {
       "default": {
         "name": "hidden_when_first",
         "type": "text",
         "label": "Hidden When First",
         "page": 0
       }
     }, {
       "default": {
         "name": "hidden_when_text_is_a",
         "type": "text",
         "label": "Hidden When Text Is A",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "maxWidthPrefix": "characters",
         "page": 0
       }
     }, {
       "default": {
         "name": "third_level",
         "type": "subForm",
         "subForm": "third_level_sub",
         "page": 0
       }
     }],
     "_checks": [{
       "default": {
         "name": "hiddenWhenFirstIsA",
         "exp": {
           "operator": "==",
           "operands": [{
             "operator": "formElement.value",
             "operands": ["hidden_when_first"]
           }, "a"]
         }
       }
     }],
     "_actions": [{
       "default": {
         "name": "hideHiddenWhenTextIsA",
         "manipulations": [{
           "target": "hidden_when_text_is_a",
           "properties": {
             "hidden": true,
             "persist": false
           }
         }]
       }
     }],
     "_behaviours": [{
       "default": {
         "name": "hideHiddenWhenTextIsA_hiddenWhenFirstIsA",
         "trigger": {
           "formElements": ["hidden_when_first"],
           "formEvents": ["load"]
         },
         "check": "hiddenWhenFirstIsA",
         "actions": [{
           "action": "hideHiddenWhenTextIsA",
           "autoReverse": true
         }]
       }
     }]
   },
   "list": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": 1,
       "view": 1,
       "delete": 1
     },
     "dataStorageList": null,
     "downloadModeList": null,
     "recordsToDisplay": null,
     "_elements": ["second_level_text", "hidden_when_first", "hidden_when_text_is_a"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "add",
     "_elements": ["second_level_text", "hidden_when_first", "hidden_when_text_is_a", "third_level"]
   },
   "edit": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "edit",
     "_elements": ["second_level_text", "hidden_when_first", "hidden_when_text_is_a", "third_level"]
   },
   "view": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "view",
     "_elements": ["second_level_text", "hidden_when_first", "hidden_when_text_is_a", "third_level"]
   },
   "delete": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "delete",
     "_elements": ["second_level_text", "hidden_when_first", "hidden_when_text_is_a", "third_level"]
   }
 }, {
   "default": {
     "uniqueNameId": "3gep",
     "name": "subform1",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "name",
         "type": "text",
         "label": "Name",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "SUBFORM1_LIST",
     "displayName": "SUBFORM1 LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["name"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "SUBFORM1_ADD",
     "displayName": "SUBFORM1 ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["name"]
   },
   "edit": {
     "interaction": "SUBFORM1_EDIT",
     "displayName": "SUBFORM1 EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["name"]
   },
   "view": {
     "interaction": "SUBFORM1_VIEW",
     "displayName": "SUBFORM1 VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["name"]
   },
   "delete": {
     "interaction": "SUBFORM1_DELETE",
     "displayName": "SUBFORM1 DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["name"]
   }
 }, {
   "default": {
     "uniqueNameId": "10v5qx2",
     "name": "sub_form_holder",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "name",
         "type": "text",
         "label": "Name",
         "page": 0
       }
     }, {
       "default": {
         "name": "required",
         "type": "text",
         "label": "Required",
         "page": 0
       }
     }, {
       "default": {
         "name": "not_required",
         "type": "text",
         "label": "Not Required",
         "page": 0
       }
     }, {
       "default": {
         "name": "inception_form",
         "type": "subForm",
         "subForm": "outer_subform",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "SUB_FORM_HOLDER_LIST",
     "displayName": "SUB_FORM_HOLDER LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["name", "required", "not_required"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "SUB_FORM_HOLDER_ADD",
     "displayName": "SUB_FORM_HOLDER ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["name", "required", "not_required", "inception_form"]
   },
   "edit": {
     "interaction": "SUB_FORM_HOLDER_EDIT",
     "displayName": "SUB_FORM_HOLDER EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["name", "required", "not_required", "inception_form"]
   },
   "view": {
     "interaction": "SUB_FORM_HOLDER_VIEW",
     "displayName": "SUB_FORM_HOLDER VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["name", "required", "not_required", "inception_form"]
   },
   "delete": {
     "interaction": "SUB_FORM_HOLDER_DELETE",
     "displayName": "SUB_FORM_HOLDER DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["name", "required", "not_required", "inception_form"]
   }
 }, {
   "default": {
     "uniqueNameId": "11g78byb",
     "name": "third_level_sub",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 4,
     "labelPlacement": null,
     "header": null,
     "footer": null,
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "third_level_req",
         "type": "text",
         "label": "Third Level Req",
         "labelPlacement": "default",
         "labelStyle": "Plain",
         "required": "1",
         "maxWidthPrefix": "characters",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": 1,
       "view": 1,
       "delete": 1
     },
     "dataStorageList": null,
     "downloadModeList": null,
     "recordsToDisplay": null,
     "_elements": ["third_level_req"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "add",
     "_elements": ["third_level_req"]
   },
   "edit": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "edit",
     "_elements": ["third_level_req"]
   },
   "view": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "view",
     "_elements": ["third_level_req"]
   },
   "delete": {
     "interaction": "",
     "displayName": "",
     "defaultCategory": "",
     "header": null,
     "footer": null,
     "hidden": "1",
     "action": "delete",
     "_elements": ["third_level_req"]
   }
 }, {
   "default": {
     "uniqueNameId": "918nr1",
     "name": "v2_sketch_pad",
     "formDescription": "",
     "defaultCategory": "",
     "maxStep": 6,
     "labelPlacement": "auto",
     "header": "",
     "footer": "",
     "_elements": [{
       "default": {
         "name": "id",
         "type": "text",
         "page": 0
       }
     }, {
       "default": {
         "name": "sketch_pad",
         "type": "draw",
         "label": "Sketch Pad",
         "size": "signature",
         "page": 0
       }
     }],
     "_checks": [],
     "_actions": [],
     "_behaviours": []
   },
   "list": {
     "interaction": "V2_SKETCH_PAD_LIST",
     "displayName": "V2_SKETCH_PAD LIST",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "list",
     "orderBy": null,
     "showLink": {
       "edit": "1",
       "view": "1",
       "delete": "1"
     },
     "dataStorageList": "temporary",
     "downloadModeList": "partial",
     "recordsToDisplay": "",
     "_elements": ["sketch_pad"]
   },
   "search": {
     "interaction": null,
     "displayName": null,
     "defaultCategory": null,
     "header": null,
     "footer": null,
     "hidden": null,
     "action": "search"
   },
   "add": {
     "interaction": "V2_SKETCH_PAD_ADD",
     "displayName": "V2_SKETCH_PAD ADD",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": null,
     "action": "add",
     "_elements": ["sketch_pad"]
   },
   "edit": {
     "interaction": "V2_SKETCH_PAD_EDIT",
     "displayName": "V2_SKETCH_PAD EDIT",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "edit",
     "_elements": ["sketch_pad"]
   },
   "view": {
     "interaction": "V2_SKETCH_PAD_VIEW",
     "displayName": "V2_SKETCH_PAD VIEW",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "view",
     "_elements": ["sketch_pad"]
   },
   "delete": {
     "interaction": "V2_SKETCH_PAD_DELETE",
     "displayName": "V2_SKETCH_PAD DELETE",
     "defaultCategory": "",
     "header": "",
     "footer": "",
     "hidden": "1",
     "action": "delete",
     "_elements": ["sketch_pad"]
   }
 }];
```

jokeyrhyme commented:
Is it worth renaming this to "WIP..." and moving the Issue back out of the Needs Review column?

ashishtilara commented:
The reason we are assigning `_elements` in subform_field's property is because of reusable subforms, you can use one subform 5 times in same form, having different fields visible/hidden each time.

simonmarklar commented:
The elements cannot be removed from the definition because they are referenced by behaviors (eg conditional logic ), and that generates an error every time the behaviors run. Behaviors are generated by the server to run on the default subform, not on the overridden subform (the properties are there for a sub form element but, assuming they can be, I have not yet found how to populate them via the builder) so i think fixing the behaviors to deal with a removed subform element is a seperate ticket.

looks like someone got a pull request in before me :O

jokeyrhyme commented:
This is fine if the only Behaviours that are broken are the ones with bad references, e.g. 1 good Behaviour and 1 broken Behaviour still allows the good one to execute. If this is the case then I'm happy to leave it. The console errors will help the form designer to debug what's going on.

If broken Behaviours break good Behaviours, then we certainly will need a separate Issue to address this.

jokeyrhyme commented:
Also, regardless of the Behaviour Issue (if there is one), I'd still really rather that the overridden Element be completely dropped from the definition, rather than have unused Backbone Models, etc floating around.

I feel that having such a Behaviour in a form definition that is going to be reused and overridden seems like a form designer problem, and not a Forms library problem.

simonmarklar commented:
merge issues fixed in another branch - https://github.com/blinkmobile/forms/issues/23
